File: TopInterview150/BinarySearch/SearchInRotatedSortedArray/solution.cpp
class Solution {
    public:
        int search(vector<int>& nums, int target) {
            if (nums[0] == target) return 0;
            if (nums.size() == 1) return -1;

            int l = 0, r = nums.size()-1;

            while (l <= r) {
                int m = l + (r-l)/2;

                if (nums[m] == target)
                {
                    return m;
                }

                if (nums[l] <= nums[m])//left half sorted
                {
                    if (nums[l] <= target && target < nums[m])//want to look left
                    {
                        r = m-1;
                    }
                    else l = m+1;
                }
                else {
                    if (nums[m] < target && target <= nums[r])
                    {
                        l = m+1;
                    }
                    else r = m-1;
                }
            }

            return -1;
        }
    };

File: TopInterview150/BinarySearch/SearchInRotatedSortedArray/solution.py
class Solution(object):
    def search(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        if nums[0] == target:
            return 0
        if len(nums) == 1:
            return -1
        
        l = 0
        r = len(nums)-1

        while l<=r:
            index = l+(r-l)/2

            if (nums[index]== target):
                return index
            
            if nums[l] <= nums[index]:
                if nums[l] <= target and target <= nums[index]:
                    r = index-1
                else:
                    l = index +1
            else:
                if nums[index] <= target and target <= nums[r]:
                    1 = index+1
                else:
                    r = index-1
        return -1